### PR TITLE
Adds support for mixed precision NVIDIA A100 Tensor Cores (F32 <= F16 * F16 + F32)

### PR DESCRIPTION
**Pull request by manishucsd**
This PR adds support for mixed precision NVIDIA A100 Tensor Core support addressing the issue #13813 from Epic #13812

F16 input for lhs, and F16 input for rhs, F32 for accumulation, and F32 output data types can be supported on CUDA backend by lowering to `mma.sync.aligned.m16n8k16.row.col.f32.f16.f16.f32` ptx instruction. Inputs and outputs are different bit-width (16b inputs and 32b outputs)

```bash
 python3 dispatch_profiler/profiler.py --dispatches=matmul_3456x1024x2048_f16t_f16t_f32t
---------------------------------------------------------------- 
Dispatch      : matmul_3456x1024x2048_f16t_f16t_f32t_tile_config_128x256_32x3_tensorcore_mmasync
Provider      : IREE Codegen
OpKind        : OperationKind.Matmul
Operation     : matmul_3456x1024x2048_f16t_f16t_f32t
Configuration : tile_config_128x256_32x3_tensorcore_mmasync
Arguments     : --batch_count=1 --m=3456 --n=1024 --k=2048 --lhs=f16t --rhs=f16t --result=f32t
                --split_k_mode=N/A --split_k_slices=N/A
Verification  : SUCCESS
Runtime(ms)   : 0.067
GFLOPs        : 216350.96
---------------------------------------------------------------- 
Dispatch      : matmul_3456x1024x2048_f16t_f16t_f32t_tile_config_128x128_64x4_tensorcore_mmasync
Provider      : IREE Codegen
OpKind        : OperationKind.Matmul
Operation     : matmul_3456x1024x2048_f16t_f16t_f32t
Configuration : tile_config_128x128_64x4_tensorcore_mmasync
Arguments     : --batch_count=1 --m=3456 --n=1024 --k=2048 --lhs=f16t --rhs=f16t --result=f32t
                --split_k_mode=N/A --split_k_slices=N/A
Verification  : SUCCESS
Runtime(ms)   : 0.069
GFLOPs        : 210079.92
---------------------------------------------------------------- 
Dispatch      : matmul_3456x1024x2048_f16t_f16t_f32t_tile_config_128x128_32x5_tensorcore_mmasync
Provider      : IREE Codegen
OpKind        : OperationKind.Matmul
Operation     : matmul_3456x1024x2048_f16t_f16t_f32t
Configuration : tile_config_128x128_32x5_tensorcore_mmasync
Arguments     : --batch_count=1 --m=3456 --n=1024 --k=2048 --lhs=f16t --rhs=f16t --result=f32t
                --split_k_mode=N/A --split_k_slices=N/A
Verification  : SUCCESS
Runtime(ms)   : 0.066
GFLOPs        : 219629.01
---------------------------------------------------------------- 
Dispatch      : matmul_3456x1024x2048_f16t_f16t_f32t_tile_config_128x64_32x5_tensorcore_mmasync
Provider      : IREE Codegen
OpKind        : OperationKind.Matmul
Operation     : matmul_3456x1024x2048_f16t_f16t_f32t
Configuration : tile_config_128x64_32x5_tensorcore_mmasync
Arguments     : --batch_count=1 --m=3456 --n=1024 --k=2048 --lhs=f16t --rhs=f16t --result=f32t
                --split_k_mode=N/A --split_k_slices=N/A
Verification  : SUCCESS
Runtime(ms)   : 0.08
GFLOPs        : 181193.93
---------------------------------------------------------------- 
Dispatch      : matmul_3456x1024x2048_f16t_f16t_f32t_tile_config_64x64_64x5_tensorcore_mmasync
Provider      : IREE Codegen
OpKind        : OperationKind.Matmul
Operation     : matmul_3456x1024x2048_f16t_f16t_f32t
Configuration : tile_config_64x64_64x5_tensorcore_mmasync
Arguments     : --batch_count=1 --m=3456 --n=1024 --k=2048 --lhs=f16t --rhs=f16t --result=f32t
                --split_k_mode=N/A --split_k_slices=N/A
Verification  : SUCCESS
Runtime(ms)   : 0.104
GFLOPs        : 139379.95
---------------------------------------------------------------- 
Dispatch      : matmul_3456x1024x2048_f16t_f16t_f32t_tile_config_64x64_32x10_tensorcore_mmasync
Provider      : IREE Codegen
OpKind        : OperationKind.Matmul
Operation     : matmul_3456x1024x2048_f16t_f16t_f32t
Configuration : tile_config_64x64_32x10_tensorcore_mmasync
Arguments     : --batch_count=1 --m=3456 --n=1024 --k=2048 --lhs=f16t --rhs=f16t --result=f32t
                --split_k_mode=N/A --split_k_slices=N/A
Verification  : SUCCESS
Runtime(ms)   : 0.105
GFLOPs        : 138052.52
```

## Top performance

For mixed-precision `F32 <= F16 * F16 + F32` is `219 TFLOPs` with `matmul_3456x1024x2048_f16t_f16t_f32t_tile_config_128x128_32x5_tensorcore_mmasync`

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: Maybe add this transformation upstream... 
Ideally we could have a way to check if a type is legal or not for a given architecture and then fold it. Not saying this is a hard requirement though, but adding this upstream is better overall.

**Review by stellaraccident — APPROVED** [review]
Nice!

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: +1 on upstreaming.
Also this is missing a test case.

Comment (grypp) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: my 2cents about upstreaming - while upstreaming offers numerous benefits, it can be time-consuming. If this transformation won't be utilized by any other piece, we can delay upstreaming to speed up development.

**Review by grypp — APPROVED** [review]
Looks clean, thanks for doing that!
don't forget adding a test

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: I have moved the folding pattern upstream and added a test, but the test is failing when run stand-alone, can someone please check, what i did incorrectly. LLVM upstream diff [here](https://github.com/iree-org/iree-llvm-fork/compare/fold_arith_ext_into_vector_contract?expand=1)

PS: I will create Phabricator patch once I get it working.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: You can combine this: `contractOp.getLhs().getDefiningOp<ExtFOp>();`

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: What is the error?

Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: If that's too hard to make it work on upstream LLVM, write a test with `iree-opt` for now.

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: In my experience upstreaming never happens. So I tend to push for adding this upstream first, and only considering putting this in iREE when it is not possible to add upstream. This is fairly hermitic, but just as a process point/convention would prefer this upstream.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
The change itself is fine... just want to make sure we make a conscious choice between adding this to IREE or MLIR

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: Please review the upstream patch [D151918](https://reviews.llvm.org/D151918). The folding in the patch is very NVIDIA A100/H100 specific and other hardware which support mixed-precision data type. 

The developmental steps to now get to #13858 (F32 <= BF16 * BF16 + F32) from #13857 (F32 <= F16 * F16  + F32): 
- Land patch [D151918](https://reviews.llvm.org/D151918) 
- IREE lvm-integrate pick up the patch 
- Rebase this PR #13857 (F32 <= F16 * F16  + F32)
- Merge this PR #13857 (F32 <= F16 * F16  + F32)
- Rebase #13858 (F32 <= BF16 * BF16 + F32) 
- Proceed with #13858 (F32 <= BF16 * BF16 + F32).

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: done

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: Does the patch D151918 look ok to land? 
Then we wait on llvm-integrate to pick this up and land this PR.

Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: Landed patch [D151918](https://reviews.llvm.org/D151918) has landed. 

Now we are going to wait for the patch to be picked up by llvm-integrate before we land this PR.

Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp]: We are now ready to merge this PR, [D151918](https://reviews.llvm.org/D151918) has landed, and we have rebased this PR. This PR now only contains IREE chagnes and no upstream changes

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**
